--- a/kernel/supercall/dispatch.c
+++ b/kernel/supercall/dispatch.c
@@ -4,6 +4,11 @@
 #include <linux/string.h>
 #include <linux/uaccess.h>
 #include <linux/version.h>
+
+#include <linux/sched.h>
+#include <linux/sched/signal.h>
+#include <linux/init_task.h>
+
 #ifdef CONFIG_KSU_SUSFS
 #include <linux/namei.h>
 #include <linux/susfs.h>
@@ -968,4 +973,4 @@
         kfree(entry);
     }
     up_write(&mount_list_lock);
-}
\ No newline at end of file
+}
