diff --git a/fs/proc/task_mmu.c b/fs/proc/task_mmu.c
index c2b18176b..5b4fdceb2 100644
--- a/fs/proc/task_mmu.c
+++ b/fs/proc/task_mmu.c
@@ -19,6 +19,10 @@
 #include <linux/shmem_fs.h>
 #include <linux/uaccess.h>
 #include <linux/mm_inline.h>
+#include <linux/ctype.h>
+#if defined(CONFIG_KSU_SUSFS_SUS_KSTAT) || defined(CONFIG_KSU_SUSFS_SUS_MAP) || defined(CONFIG_KSU_SUSFS_OPEN_REDIRECT)
+#include <linux/susfs_def.h>
+#endif // #if defined(CONFIG_KSU_SUSFS_SUS_KSTAT) || defined(CONFIG_KSU_SUSFS_SUS_MAP) || defined(CONFIG_KSU_SUSFS_OPEN_REDIRECT)
 
 #include <asm/elf.h>
 #include <asm/tlb.h>

diff --git a/fs2/namei.c b/fs/namei.c
index 0120e4b..d8fa21c 100644
--- a/fs2/namei.c
+++ b/fs/namei.c
@@ -1710,24 +1710,23 @@ retry:
 		return ERR_PTR(-ENOMEM);
 
 #ifdef CONFIG_KSU_SUSFS_SUS_PATH
-	dentry = lookup_real(base->d_inode, dentry, flags);
-
-	if (unlikely(dentry) && !IS_ERR(dentry) && dentry->d_inode && !found_sus_path && susfs_is_inode_sus_path(dentry->d_inode)) {
-		if (d_in_lookup(dentry))
-			d_lookup_done(dentry);
-		if (!(flags & LOOKUP_RCU))
-			dput(dentry);
-		// - Just in case if an user app has been granted full file access and
-		//   it is trying to find the fuse sus path with the create flag, then
-		//   at least we can prevent the fake qstr file from from being created,
-		//   although it is futile to do this, it is better than doing nothing.
-		if (dentry->d_inode->i_sb->s_magic == FUSE_SUPER_MAGIC &&
-			(flags & (LOOKUP_CREATE | LOOKUP_EXCL)))
-			return ERR_PTR(-EACCES);
-		dentry = d_alloc(base, &susfs_fake_qstr_name);
-		found_sus_path = true;
-		goto retry;
-	}
+    dentry = lookup_real(base->d_inode, dentry, flags);
+    if (unlikely(dentry) && !IS_ERR(dentry) && dentry->d_inode && !found_sus_path && susfs_is_inode_sus_path(dentry->d_inode)) {
+        if (d_in_lookup(dentry))
+            d_lookup_done(dentry);
+            
+        bool is_fuse_create = (dentry->d_inode->i_sb->s_magic == FUSE_SUPER_MAGIC && (flags & (LOOKUP_CREATE | LOOKUP_EXCL)));
+
+        if (!(flags & LOOKUP_RCU))
+            dput(dentry);
+
+        if (is_fuse_create)
+            return ERR_PTR(-EACCES);
+
+        dentry = d_alloc(base, &susfs_fake_qstr_name);
+        found_sus_path = true;
+        goto retry;
+    }
 #endif
 
 #ifdef CONFIG_KSU_SUSFS_SUS_PATH

diff --git a/fs/statfs.c b/fs/statfs.c
index fa95afe..da2d69a 100644
--- a/fs/statfs.c
+++ b/fs/statfs.c
@@ -96,21 +96,22 @@ int vfs_statfs(const struct path *path, struct kstatfs *buf)
 #endif // #ifdef CONFIG_KSU_SUSFS_OPEN_REDIRECT
 
 #ifdef CONFIG_KSU_SUSFS_SUS_MOUNT
-	if (likely(susfs_is_current_proc_umounted() && path->mnt)) {
-		no_sus_vfsmnt = susfs_get_non_sus_vfsmnt_from_vfsmnt(path->mnt);
-		if (path->mnt == no_sus_vfsmnt) {
-			dput(no_sus_vfsmnt->mnt_root);
-			mntput(no_sus_vfsmnt);
-			goto orig_flow;
-	}
-		error = statfs_by_dentry(no_sus_vfsmnt->mnt_root, buf);
-		if (!error)
-			buf->f_flags = calculate_f_flags(no_sus_vfsmnt);
-		dput(no_sus_vfsmnt->mnt_root);
-		mntput(no_sus_vfsmnt);
-		return error;
-	}
-#endif // #ifdef CONFIG_KSU_SUSFS_SUS_MOUNT
+    if (likely(susfs_is_current_proc_umounted() && path->mnt)) {
+        no_sus_vfsmnt = susfs_get_non_sus_vfsmnt_from_vfsmnt(path->mnt);
+        
+        if (path->mnt == no_sus_vfsmnt) {
+            mntput(no_sus_vfsmnt);
+            goto orig_flow;
+        }
+        
+        error = statfs_by_dentry(no_sus_vfsmnt->mnt_root, buf);
+        if (!error)
+            buf->f_flags = calculate_f_flags(no_sus_vfsmnt);
+            
+        mntput(no_sus_vfsmnt);
+        return error;
+    }
+#endif
 
 #if defined(CONFIG_KSU_SUSFS_SUS_MOUNT) || defined(CONFIG_KSU_SUSFS_OPEN_REDIRECT)
 orig_flow:

diff --git a/fs/susfs.c b/fs/susfs.c
index 69c66c8..dc87c1d 100644
--- a/fs/susfs.c
+++ b/fs/susfs.c
@@ -990,13 +990,15 @@ int susfs_open_redirect_spoof_vfs_readlink(struct inode *inode, char __user *buf
 				srcu_read_unlock(&susfs_srcu_open_redirect, srcu_idx);
 				return -ENAMETOOLONG;
 			}
-			if (copy_to_user(buffer, entry->info.redirected_pathname, strlen(entry->info.redirected_pathname))) {
-				SUSFS_LOGE("copy_to_user() failed\n");
-				srcu_read_unlock(&susfs_srcu_open_redirect, srcu_idx);
-				return -EFAULT;
-			}
-			srcu_read_unlock(&susfs_srcu_open_redirect, srcu_idx);
-			return 0;
+			
+			int len = strlen(entry->info.redirected_pathname);
+		        if (copy_to_user(buffer, entry->info.redirected_pathname, len)) {
+		                SUSFS_LOGE("copy_to_user() failed\n");
+		                srcu_read_unlock(&susfs_srcu_open_redirect, srcu_idx);
+		                return -EFAULT;
+		        }
+		        srcu_read_unlock(&susfs_srcu_open_redirect, srcu_idx);
+		        return len;
 		}
 	}
 	srcu_read_unlock(&susfs_srcu_open_redirect, srcu_idx);
@@ -1359,10 +1361,10 @@ static int watch_one_dir(struct watch_dir *wd)
  */
 static SUSFS_DECL_FSNOTIFY_OPS(susfs_handle_sdcard_inode_event)
 {
-	if (!file_name || strlen((const char *)file_name) != 7 ||
-	    memcmp(file_name, "Android", 7))
-		return 0;
-
+        if (!file_name || susfs_fname_len(file_name) != 7 ||
+            memcmp(susfs_fname_arg(file_name), "Android", 7))
+                return 0;
+                
 	if (test_and_set_bit(0, &sdcard_cleanup_scheduled))
 		return 0;
 
